### PR TITLE
User data change

### LINKED ISSUE #285: Change data schema
Right now we have 4 fields, `posRep`, `negRep`, `graffiti`, `timestamp`. I propose we change our data structure to 6 256 bit** fields:

- `data[0]` - combined with addition
- `data[1]` - combined with addition
- `data[2]` - combined with addition
- `data[3]` - combined with addition
- `data[4]` - combined by replacement
- `data[5]` - `data[4]` metadata field (timestamp, replacement number)

Showing user data as raw bit fields might make it easier for developers to conceptualize the system. Addition fields are cheap to add in ZK, each addition field adds < 1 poseidon hash to the ust proof.

Applications can split each 256 bit structure using additions with shifted bits. For example, say I'm using an attester that allows me to anonymously provide liquidity to a uniswap pool. In `data[0]` I could store the following:
- `data[0]`: 128 bits liquidity ownership
- `data[0]`: 64 bits deposit timestamp
- `data[0]`: 64 bits accrued fees

To set the deposit timestamp I would attest giving the user `timestamp << 128`. This will cause the 64 bits at bit offset 128 to be added to. Each field needs to be large enough that overflows are impossible. It's difficult for the user to prove their balance to prevent an overflow so the field must be large enough that it won't happen under normal operating conditions.

A more safe schema using 2 data fields might be:
- `data[0]`: 128 bits liquidity ownership
- `data[0]`: 128 bits accrued fees
- `data[1]`: 64 bits timestamp

Adding more data fields is a simple way to make the system more usable for applications.

**: Each field is technically not 256 bits. They are the size of the snark field, which is ~2^254, so 254 bits.

### LINKED ISSUE #306: `firstLeafPreImage` and `lastLeafPreImage`
https://github.com/Unirep/Unirep/pull/282#discussion_r1097555114

### LINKED ISSUE #321: Change epoch leaf structure
We store epoch tree leaves as `H(epochKey, posRep, negRep, graffiti, timestamp)`. The problem with this is each `submitAttestation` must do a `Poseidon5()`, which is at least 100k gas.

Instead we can calculate an epoch tree using a polysum.

`P(H(epochKey), H(posRep), H(negRep), H(graffiti), H(timestamp))`

This could also be written like (calling `posRep`/`negRep`/`graffiti`/`timestamp` $d_{0-3}$). $R$ is a random constant.

$leaf = H(epochKey)*R + H(d_0)*R^2 + H(d_1)*R^3 + H(d_2)*R^4 + H(d_3)*R^5$

The important thing here is, if we're changing only `posRep`, instead of doing `Poseidon5` we do `Poseidon1` and then a couple modular addition/subtractions. Poseidon1 costs ~18k gas vs the 100k for `Poseidon5`.

This is also good because we can add more data fields without needing a larger poseidon onchain.

#### First attestation

The first time a user receives an attestation we'll need to do two `Poseidon1` hashes. First we hash the epoch key, and then we hash the data field (e.g. `posRep`). We'll also assign all other fields `H(0)`.

We need to hash the epoch key and set the unused fields to `H(0)` so we can prove the pre-image in the `buildOrderedTree` proof. To safely use the Polysum we have to prove all the pre-images of inputs in ZK.

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #321 
Closes #285 
Closes #306

## What is the new behavior?

Adds two new configuration variables: `FIELD_COUNT`, `SUM_FIELD_COUNT`. 

`FIELD_COUNT` is the total number of data fields the user has.
`SUM_FIELD_COUNT` is the number of data fields that are combined using addition.
The remaining fields are combined using replacement. Each replacement field uses two indexes for storing data and timestamp.

So if we have `FIELD_COUNT = 4` and `SUM_FIELD_COUNT = 2` we have 2 addition fields and 1 replacement field. (e.g. `posRep`/`negRep`/`graffiti`)

--

Refactors epoch tree and state tree leaves to use `Poseidon1` inside a polysum. This is slightly more expensive but allows us to scale to more than 16 `FIELD_COUNT`. It's also cheaper onchain to make attestations.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

File: packages/core/test/epochTree.test.ts [hunk 1]
```diff
@@ -110,37 +111,32 @@ describe('Epoch tree', function () {
...
                     .then((t) => t.wait())
                     .then(({ blockNumber }) =>
                         ethers.provider.getBlock(blockNumber)
                     )
-                posRep += newPosRep
-                negRep += newNegRep
-                graffiti = newGraffiti
-                timestamp = newTimestamp
+                if (fieldIndex < userState.sync.settings.sumFieldCount) {
+                    data[fieldIndex] += BigInt(val)
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/core/test/epochTree.test.ts:133]: i think it's unnecessary because the sum fields use modular addition in the field like [here](https://github.com/Unirep/Unirep/pull/322/files#diff-bd4415896551787d4d50aa8e8bf1cf39f56c05251c9f8fdb9b1746aca509ec02R249)

File: packages/circuits/circuits/proveReputation.circom [hunk 2]
```diff
@@ -81,33 +80,40 @@ template ProveReputation(STATE_TREE_DEPTH, EPOCH_KEY_NONCE_PER_EPOCH, MAX_REPUTA
...
     /* 2. Check if user has reputation greater than min_rep */
-    // if proving min_rep > 0, check if pos_rep >= neg_rep + min_rep
+    // if proving min_rep > 0, check if data[0] >= data[1] + min_rep
 
-    component min_rep_check = GreaterEqThan(MAX_REPUTATION_SCORE_BITS);
-    min_rep_check.in[0] <== pos_rep;
-    min_rep_check.in[1] <== neg_rep + min_rep;
+    component min_rep_check = GreaterEqThan(252);
+    min_rep_check.in[0] <== data[0];
+    min_rep_check.in[1] <== data[1] + min_rep;
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/circuits/circuits/proveReputation.circom:116]: I think this is pretty well abstracted by the `core` package. We really need a new proof allowing the user to select which fields they want to prove, but that seems out of scope for this pr

File: packages/contracts/contracts/interfaces/IUnirep.sol [hunk 3]
```diff
@@ -68,6 +62,7 @@ interface IUnirep {
     error NullifierAlreadyUsed(uint256 nullilier);
     error AttesterIdNotMatch(uint160 attesterId);
     error OutOfRange();
+    error InvalidField();
```
Comment (vivianjeng) [inline packages/contracts/contracts/interfaces/IUnirep.sol:65]: should update documentation
you can run `yarn contracts errorcode` to see the error name and its sigHash

File: packages/core/test/epochTree.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply packages/core/test/epochTree.test.ts:133]: yeah it is handled on-chain
but I think generating a test that does not handle edge cases is not good
it can be like UserState
```ts
    data[fieldIndex] = data[fieldIndex] + (val % F)
```

File: packages/circuits/circuits/proveReputation.circom (hunk 2 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply packages/circuits/circuits/proveReputation.circom:116]: we can note that the `data[0]` is from pos_rep and `data[1]` is from `neg_rep` by comment.
or after this PR is merged, no `pos_rep` and `neg_rep` is used

File: packages/circuits/test/proveReputation.test.ts [hunk 4]
```diff
@@ -429,12 +430,7 @@ describe('Prove reputation from attester circuit', function () {
             epoch,
             nonce,
             attesterId,
-            startBalance: {
-                posRep: 0,
-                negRep: 0,
-                graffiti: 100191,
-                timestamp: 0,
-            },
+            startBalance: [...Array(SUM_FIELD_COUNT).fill(0), 100191],
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/circuits/test/proveReputation.test.ts:433]: This happens in the circuits `test/utils`

File: packages/circuits/src/ReputationProof.ts [hunk 5]
```diff
@@ -13,6 +13,7 @@ export class ReputationProof extends BaseProof {
         control0: 2,
         control1: 3,
         graffitiPreImage: 4,
+        data: 5,
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/circuits/src/ReputationProof.ts:16]: It's tested in `epochKeyLite` the other circuits re-use this component

File: packages/circuits/test/proveReputation.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply packages/circuits/test/proveReputation.test.ts:433]: I mean the `genReputationCircuitInput` can have inputs `sumField` and `replField`
and in that function it concatenates two arrays into a `data` array

e.g. in `proveReputation.test.ts`
```ts
        const circuitInputs = genReputationCircuitInput({
            ...
            sumField: [5, 10],
            replField: [100191],
            ...
        })
```
and in `test/utils`
```ts
const startBalance = [
        ...sumField,
        ...Array(SUM_FIELD_COUNT - sumField.length).fill(0),
        ...replField,
        ...Array(FIELD_COUNT - SUM_FIELD_COUNT - replField.length).fill(0),
    ]
```

File: packages/circuits/src/ReputationProof.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (vivianjeng) [reply packages/circuits/src/ReputationProof.ts:16]: then in `core/UserState`, it should include `data` in `genProveReputationProof`
https://github.com/Unirep/Unirep/blob/d25111f7f2fd52f9b5d5982aa88beb0f78ddb752/packages/core/src/UserState.ts#L481-L488

File: packages/circuits/test/proveReputation.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/circuits/test/proveReputation.test.ts:433]: Ok I understand, but I think this can be changed later.

File: packages/circuits/src/ReputationProof.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/circuits/src/ReputationProof.ts:16]: [fe007d9](https://github.com/Unirep/Unirep/pull/322/commits/fe007d9557d348686c76359f31c5618dd55326e2)

File: packages/core/src/UserState.ts [hunk 6]
```diff
@@ -286,14 +271,15 @@ export default class UserState {
...
         })
         for (const a of attestations) {
-            posRep += BigInt(a.posRep)
-            negRep += BigInt(a.negRep)
-            if (a.timestamp && BigInt(a.timestamp) > timestamp) {
-                graffiti = BigInt(a.graffiti)
-                timestamp = BigInt(a.timestamp)
+            const { fieldIndex } = a
+            if (fieldIndex < this.sync.settings.sumFieldCount) {
+                data[fieldIndex] = data[fieldIndex] + (BigInt(a.change) % F)
```
[the remark this answers is not in the document]
Comment (vimwitch) [reply packages/core/src/UserState.ts]: Good catch, fixed and tested here [680958d](https://github.com/Unirep/Unirep/pull/322/commits/680958d0d74e78bd69571af5a5b84674d63fc61f)

This logic is well tested in the contracts here [a6a4e61](https://github.com/Unirep/Unirep/pull/322/commits/a6a4e61b980f52fbee5ea16e561b09bb5ee95e8d)

**Review by vivianjeng — APPROVED** [review]
